src/main/java/org/spongepowered/api/text/TextTemplate.java

```diff
@@ -0,0 +1,473 @@
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import javax.annotation.Nullable;
+
+/**
+ * Represents a re-usable template that produces a formatted
+ * {@link Text.Builder}. Elements will be appended to the result builder in the
+ * order that they are specified in {@link #of(Object...)}.
+ */
+public class TextTemplate implements TextRepresentable {
+
+    static {
+        TypeSerializers.getDefaultSerializers().registerType(TypeToken.of(TextTemplate.class), new TextTemplateConfigSerializer());
+    }
+
+    /**
+     * Default "open" String for how arguments are contained within the template.
+     */
+    public static final String DEFAULT_OPEN_ARG = "{";
+
+    /**
+     * Default "close" String for how arguments are contained within the template.
+     */
+    public static final String DEFAULT_CLOSE_ARG = "}";
+
+    /**
+     * Empty representation of a {@link TextTemplate}. This is returned if the
+     * array supplied to {@link #of(Object...)} is empty.
+     */
+    public static final TextTemplate EMPTY = new TextTemplate(DEFAULT_OPEN_ARG, DEFAULT_CLOSE_ARG, new Object[]{});
+
+    protected final List<Object> elements = new ArrayList<>();
+    protected final Map<String, Arg> arguments = new HashMap<>();
+    protected final String openArg, closeArg;
+
+    protected TextTemplate(String openArg, String closeArg, Object[] elements) {
+        this.openArg = openArg;
+        this.closeArg = closeArg;
+        for (Object element : elements) {
+            if (element instanceof Arg.Builder) {
+                element = ((Arg.Builder) element).build();
+            }
+            this.elements.add(element);
+            if (element instanceof Arg) {
+                // check for non-equal duplicate argument
+                Arg newArg = (Arg) element;
+                newArg.openArg = this.openArg;
+                newArg.closeArg = this.closeArg;
+                Arg oldArg = this.arguments.get(newArg.name);
+                if (oldArg != null && !oldArg.equals(newArg)) {
+                    throw new TextTemplateArgumentException("Tried to supply an unequal argument with a duplicate name \""
+                            + newArg.name + "\" to TextTemplate.");
+                }
+                arguments.put(newArg.name, newArg);
+            }
+        }
+    }
+
+    /**
+     * Returns the elements contained in this TextTemplate.
+     *
+     * @return The elements within the template
+     */
+    public List<Object> getElements() {
+        return ImmutableList.copyOf(this.elements);
+    }
+
+    /**
+     * Returns the arguments contained within the TextTemplate.
+     *
+     * @return The arguments within this TextTemplate
+     */
+    public Map<String, Arg> getArguments() {
+        return ImmutableBiMap.copyOf(this.arguments);
```

Comment:
...what? The original is not a `BiMap`. I assume you meant `ImmutableMap.copyOf(this.arguments)`?